records/2024-06-04.manifests-kustomize-cicd.md

```diff
@@ -0,0 +1,140 @@
+
+#### Diagram
+
+![Production Current CICD](./diagrams/2024-06-04.manifests-current-cicd/production-cicd.png)
+
+### Problems
+
+#### Two Different Deployment Technologies
+
+We are currently managing two different Kubernetes deployment methods with Kustomize and Helmfile. It would be beneficial to have everything under one technology.
+
+#### Staging Releases
+
+Because we do not actually update any manifests code when doing an individual application rollout, we lose all context of which version we should be using when we do a global manifest rollout with merge to main (staging). To overcome this, the staging Kustomize code is set to always use the "latest" tag. This means when we apply kustomize the explicitly versioned docker tag is overwritten with "latest". This works because the docker build and push job for each application tags as both latest and the github sha. It is however not ideal since it becomes difficult to track versions in staging, and also creates a delta between how staging and production releases are handled through release pipelines.
+
+This issue is a problem regardless of whether we use Kustomize or Helmfile.
+
+#### Helmfile Image Versions
+
+For helmfile releases, we are relying on the helmfile logic to set the image versions in the overrides. For example, we may have something like the code block below, which hard codes the tag for production, but accepts either an environment variable or will default to "latest" if it does not exist.
+
+```yaml
+      {{ if eq .Environment.Name "production" }}
+      image:
+        tag: release1.24
+      {{ else }}
+      image:
+        tag: {{ env "DOCKER_TAG" | default "latest" }}
+      {{ end }}
+```
+
+This is an ok solution, but it's not ideal to have to modify the individual override files to set image versions for production manually. It would be ideal to have this done automatically like with the PR bot.
+
+## Options
+
+Because Helmfile has been identified as a long term goal to migrate GC Notify to, the options below will all involve moving the Kustomize code to Helmfile.
+
+### Option 1: Migrate Everything to Helmfile Keeping the Existing CI/CD Processes
+
+The first option is to simply move Notify to helmfile, and adapt the existing CI/CD process. This will be a relatively easy change as most of this work has been completed in a manifests branch. The Notify PR bot will have to be modified to change the image tag in the overrides files of helmfile instead of the Kustomization.yaml file.
+
+The staging deployments would remain exactly the same as they are now.
+
+#### Diagram
+
+![Production CICD Option 1](./diagrams/2024-06-04.manifests-current-cicd/cicd-option1.png)
+
+#### Pros
+
+- Least effort
+- Consolidates all Kubernetes resources to a single technology
+- Introduces the ability to add "Diff" previews to Notify Applications on manifests PRs
```

Comment:
```suggestion
- Introduces the ability to add "Diff" previews to Notify Applications on manifests PRs
- Incremental (toward other options)
- Focus and isolate Helm migration
```

or something along these line? we can do this as a first step and still move past this option afterward. The Helm migration is already a big one by itself so it might be better to isolate changes from this migration first and focus properly.